File: Tema 4. Árboles/Prácticas/MonticulosMaxiphobic/Maxiphobic.haskell/MaxiphobicHeap.hs
-------------------------------------------------------------------------------
-- Estructuras de Datos. 2º Curso. ETSI Informática. UMA
--
-- PRACTICA 6. Montículos maxifóbicos en Haskel
--
-- Alumno: Ángel Manuel Soria Gil
-------------------------------------------------------------------------------

-- Solo tienes que completar el código de la función 'merge' de acuerdo con
-- las explicaciones del enunciado.

module MaxiphobicHeap
  ( Heap
  , empty
  , isEmpty
  , minElem
  , delMin
  , insert
  , merge
-- los siguientes son auxiliares
  , mkHeap
  , verifyOP
  , drawOnWith
  ) where

import           DataStructures.Graphics.DrawTrees
import           Test.QuickCheck

data Heap a = Empty | Node a Int (Heap a) (Heap a) deriving Show

size :: Heap a -> Int
size Empty           = 0
size (Node _ sz _ _) = sz

empty :: Heap a
empty  = Empty

isEmpty :: Heap a -> Bool
isEmpty Empty = True
isEmpty _     = False

singleton :: a -> Heap a
singleton x  = Node x 1 Empty Empty

insert :: (Ord a) => a -> Heap a -> Heap a
insert x h  = merge h (singleton x)

minElem :: Heap a -> a
minElem Empty          = error "minElem on empty heap"
minElem (Node x _ _ _) = x

delMin :: (Ord a) => Heap a -> Heap a
delMin Empty            = error "delMin on empty heap"
delMin (Node _ _ lh rh) = merge lh rh

----------------------------------------------------------
-- VVVVVVVVVVVV-SOLO TOCAR ABAJO-VVVVVVVVVVVV-------------
----------------------------------------------------------

{-
    Al mezclar tenemos tres montículos: el ganador y los hijos del perdedor.
    Se cuelga a la izquierda el más pesado de los tres y a la derecha la mezcla
    de los más ligeros:

                           mínimo
                          /      \
                       más       mezcla de los
                     pesado       más ligeros

   Política de desempate:

   Si hay empate para elegir el montículo más pesado:

    1. elegir preferentemente al hijo izquierdo del perdedor
    2. si no, elegir preferentemente al hijo derecho del perdedor
    3. si no, elegir al ganador
-}

-- recursively merges smallest subheaps. Achieves O(log n) complexity
merge :: (Ord a) => Heap a -> Heap a -> Heap a
merge m1 Empty = m1
merge Empty m1 = m1
merge h1@(Node x1 s1 lt1 rt1) h2@(Node x2 s2 lt2 rt2) -- El @ se usa para ponerle un nombre a cada árbol con el que referirnos
  | x1 <= x2  = Node x1 (s1+s2) masPesado1 (merge o1 o2)
  | otherwise = Node x2 (s1+s2) masPesado2 (merge o3 o4)
  where 
    (masPesado1, o1, o2) = ordena h2 lt1 rt1
    (masPesado2, o3, o4) = ordena h1 lt2 rt2
    ordena w i d
      | size w > size i && size w > size d = (w,i,d)
      | size i >= size w && size i >= size d = (i,w,d)
      | size d >= size w && size d > size i = (d,w,i)
    
 
----------------------------------------------------------
-- ^^^^^^^^^^^^^^-- SOLO TOCAR ARRIBA ^^^^^^^^^^^ --------
----------------------------------------------------------

-- Efficient O(n) bottom-up construction for heaps
mkHeap :: (Ord a) => [a] -> Heap a
mkHeap []  = empty
mkHeap xs  = mergeLoop (map singleton xs)
  where
    mergeLoop [h] = h
    mergeLoop hs  = mergeLoop (mergePairs hs)

    mergePairs []        = []
    mergePairs [h]       = [h]
    mergePairs (h:h':hs) = merge h h' : mergePairs hs

-------------------------------------------------------------------------------
-- Generating arbitrary Heaps
-------------------------------------------------------------------------------

instance (Ord a, Arbitrary a) => Arbitrary (Heap a) where
  arbitrary  = do
    xs <- arbitrary
    return (mkHeap xs)

-------------------------------------------------------------------------------
-- Invariants
-------------------------------------------------------------------------------

verifyOP :: (Ord a) => Heap a -> Bool
verifyOP Empty             = True
verifyOP (Node x _ lh rh)  = x `lessEq` lh && x `lessEq` rh
                           && verifyOP lh && verifyOP rh
 where
  x `lessEq` Empty           = True
  x `lessEq` (Node x' _ _ _) = x<=x'

-------------------------------------------------------------------------------
-- Drawing a Heap
-------------------------------------------------------------------------------

instance Subtrees (Heap a) where
  subtrees Empty            = []
  subtrees (Node _ _ lh rh) = [lh,rh]

  isEmptyTree Empty = True
  isEmptyTree _     = False

instance (Show a) => ShowNode (Heap a) where
  showNode (Node x _ _ _) = show x

drawOnWith :: FilePath -> (a -> String) -> Heap a -> IO ()
drawOnWith file toString = _drawOnWith file showHeap
 where
  showHeap (Node x _ _ _) = toString x

File: Tema 4. Árboles/Prácticas/MonticulosMaxiphobic/prMaxiphobic/src/dataStructures/heap/MaxiphobicHeap.java
/*
 * Estructuras de Datos. 2º Curso. ETSI Informática. UMA
 * PRACTICA 6. Montículos maxifóbicos en Java
 * APELLIDOS, NOMBRE:
 */

package dataStructures.heap;

// Solo tienes que completar el método merge.

/**
 * Heap implemented using maxiphobic heap-ordered binary trees.
 *
 * @param <T> Type of elements in heap.
 */
public class MaxiphobicHeap<T extends Comparable<? super T>> implements Heap<T> {

    private static class Tree<E> {
        private final E elem;
        private final int size;
        private final Tree<E> left;
        private final Tree<E> right;

        public Tree(E x, Tree<E> l, Tree<E> r) {
            elem = x;
            left = l;
            right = r;
            size = 1 + size(l) + size(r);
        }
    }

    private Tree<T> root;

    /**
     * Creates an empty Maxiphobic Heap.
     * <p>
     * Time complexity: O(1)
     */
    public MaxiphobicHeap() {
        root = null;
    }

    private static int size(Tree<?> heap) {
        return heap == null ? 0 : heap.size;
    }

    private static boolean isLeaf(Tree<?> current) {
        return current != null && current.left == null && current.right == null;
    }

    /*
        Al mezclar tenemos tres montículos: el ganador y los hijos del perdedor.
        Se cuelga a la izquierda el más pesado de los tres y a la derecha la mezcla
        de los más ligeros:

                               mínimo
                              /      \
                           más       mezcla de los
                         pesado       más ligeros

       Política de desempate:

       Si hay empate para elegir el montículo más pesado:

        1. elegir preferentemente al hijo izquierdo del perdedor
        2. si no, elegir preferentemente al hijo derecho del perdedor
        3. si no, elegir al ganador
    */


    private static <T extends Comparable<? super T>> Tree<T> merge(Tree<T> h1, Tree<T> h2) {
        if (h1.size == 0 && h2.size == 0){  // Caso base con ambos nulos
            return null;
        } else if (h1.size == 0){           // Caso base con izquierdo nulo
            return h2;
        } else if (h2.size == 0){           // Caso base con derecho nulo
            return h1;
        } else {                            // Ninguno nulo
            T el;
            Tree<T> l;
            Tree<T> r;
            if (h1.elem.compareTo(h2.elem)<=0){ //el primero pierde
                el = h1.elem;                   //seleccionamos el mínimo
                if (h1.left.size>=h1.right.size && h1.left.size>= h2.size){
                    l = h1.left;
                    r = merge(h1.right, h2);
                } else if (h1.right.size>h1.left.size && h1.right.size>=h2.size) {
                    l = h1.right;
                    r = merge(h1.left, h2);
                } else {
                    l = h2;
                    r = merge(h1.left, h1.right);
                }
            } else { // el primero gana
                el = h2.elem; //seleccionamos el mínimo
                if (h2.left.size>=h2.right.size && h2.left.size>= h1.size){
                    l = h2.left;
                    r = merge(h2.right, h1);
                } else if (h2.right.size>h2.left.size && h2.right.size>=h1.size) {
                    l = h2.right;
                    r = merge(h2.left, h1);
                } else {
                    l = h1;
                    r = merge(h2.left, h2.right);
                }
            }
            return new Tree<>(el, l, r);
        }
    }

    /**
     * {@inheritDoc}
     * <p>
     * Time complexity: O(1)
     */
    @Override
    public boolean isEmpty() {
        return root == null;
    }

    /**
     * {@inheritDoc}
     * <p>
     * Time complexity: O(1)
     */
    @Override
    public int size() {
        return root == null ? 0 : root.size;
    }

    /**
     * {@inheritDoc}
     * <p>
     * Time complexity: O(1)
     *
     * @throws <code>EmptyHeapException</code> if heap stores no element.
     */
    @Override
    public T minElem() {
        if (isEmpty())
            throw new EmptyHeapException("minElem on empty heap");
        else
            return root.elem;
    }

    /**
     * {@inheritDoc}
     * <p>
     * Time complexity: O(log n)
     *
     * @throws <code>EmptyHeapException</code> if heap stores no element.
     */
    @Override
    public void delMin() {
        if (isEmpty())
            throw new EmptyHeapException("delMin on empty heap");
        else
            root = merge(root.left, root.right);
    }

    /**
     * {@inheritDoc}
     * <p>
     * Time complexity: O(log n)
     */
    @Override
    public void insert(T value) {
        Tree<T> newHeap = new Tree<>(value, null, null);
        root = merge(root, newHeap);
    }

    public void clear() {
        root = null;
    }

    /**
     * Returns representation of tree as a String.
     */
    @Override
    public String toString() {
        return getClass().getSimpleName() + "(" + toStringRec(this.root) + ")";
    }

    /**
     * Returns a String with the representation of tree in DOT (graphviz).
     */
    public String toDot(String treeName) {
        final StringBuffer sb = new StringBuffer();
        sb.append(String.format("digraph \"%s\" {\n", treeName));
        sb.append("node [fontname=\"Arial\", fontcolor=red, shape=circle, style=filled, color=\"#66B268\", fillcolor=\"#AFF4AF\" ];\n");
        sb.append("edge [color = \"#0070BF\"];\n");
        toDotRec(root, sb);
        sb.append("}");
        return sb.toString();
    }

    private static String toStringRec(Tree<?> tree) {
        return tree == null ? "null" : "Node<" + toStringRec(tree.left) + ","
                + tree.elem + "," + toStringRec(tree.right)
                + ">";
    }

    private static void toDotRec(Tree<?> current, StringBuffer sb) {
        if (current != null) {
            final int currentId = System.identityHashCode(current);
            sb.append(String.format("%d [label=\"%s\"];\n", currentId, current.elem));
            if (!isLeaf(current)) {
                processChild(current.left, sb, currentId);
                processChild(current.right, sb, currentId);
            }
        }
    }

    private static void processChild(Tree<?> child, StringBuffer sb, int parentId) {
        if (child == null) {
            sb.append(String.format("l%d [style=invis];\n", parentId));
            sb.append(String.format("%d -> l%d;\n", parentId, parentId));
        } else {
            sb.append(String.format("%d -> %d;\n", parentId, System.identityHashCode(child)));
            toDotRec(child, sb);
        }
    }
}
